building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+
+- Add `sse.hpp` and `sse.cpp` in subfolder `sse`
+
+`sse.cpp`:
+
+```cpp
+#include "sse.hpp"
+#include <iostream>
+
+void sse()
+{
+  std::cout << "SSE is great!" << std::endl;
+}
+```
+
+`sse.hpp`:
+
+```cpp
+#pragma once
+
+void sse();
+```
+
+`main.cpp`:
+
+```diff
+#include <iostream>
++ #include "sse.hpp"
+
+int main()
+{
+  std::cout << "Hello World!" << std::endl;
++ sse();
+  return 0;
+}
+```
+
+`Makefile`:
+
+```makefile
+sse.o : sse/sse.hpp sse/sse.cpp
+	g++ -c sse/sse.cpp
+
+HelloWorld : main.cpp sse/sse.hpp
+	g++ -o HelloWorld sse.o main.cpp sse.o
+```
+
+- Run `make`, only builds `sse.o`
+- By default, first target is built
+- `make HelloWorld` to build specific target
+- phony target (doesn't correspond to file, no update rule)
+
+```diff
++ all : HelloWorld sse.o
+```
+
+- Why does Make not just build directly build all targets? We could want to do different things with different targets.
+- Add `clean` target, has no dependency
+
+```diff
++ clean :
++ 	rm HelloWorld sse.o
+```
+
+- Run `make clean`
+- `mkdir clean` and `make clean` confuses Make
+
+```
++ .PHONY : all clean
+```
+
+### Advanced Make
+
+- There is more:
+    - Variables
+    - Rules
+    - Wildcards
+- ... but becomes quickly very hard to read
+- Not covered because cmake does this for us.
+- But nicely documented in [py-RSE chapter 9](https://merely-useful.tech/py-rse/automate.html).
+
+
+## CMake
+
+*35 mins* ... might be tough
+
+### Introduction
+
+
+What is cmake?
+
+> A build system generator, for e.g. makefiles
+
+Why not directly use makefiles?
+
+> CMake can do many more things, is more flexible, avoids complex patterns, and is, thus, more readable
+
+Are there alternatives?
+
+> Yes, `scons`, `autotools`, ...
+
+Will you explain and discuss the differences in detail?
+
+> No, but I claim that CMake is the most standard tool and sticking to standards has a value in itself. [Watch a video why](https://youtu.be/sBP17HQAQjk)
+
+Can you give some reasons why CMake is great?
+
+> Of course:
+
+- CMake can generate many build systems: make, ninja, VSCode project, Eclipse project, ...
+- Many GUIs and TUIs: ccmake, cmake-gui, integration in probably nearly all IDEs, ...
+- CMake is *"cross-platform"*: you can ideally use same cmake file in all OS's (easy to distinguish platform-specific things)
+- Build directory becomes independent of source directory (build multiple versions with different dependencies etc.)
+- CMake respects choices in user environment (e.g. user defines cpp compiler through CXX)
+
+### CMake "Hello World"
+
+Example adapted from [Yanson Tech](https://www.youtube.com/watch?v=wl2Srog-j7I)
+Go back to original `main.cpp`
+
+`CMakeLists.txt`:
+
+```cmake
+cmake_minimum_required(VERSION "3.12")
+
+project("HelloWorld")
+
+add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- CMake variable values: `${}`, use quotations to avoid problems with spaces
+- `mkdir build && cd build && cmake ..` ... explain files (TODO), look at makefile
```

Comment:
- Explain how to pass parameters and that one always has to prefix them with `-D`?
- Do you explain default [CMake variables and their values](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html) such as `CMAKE_BUILD_TYPE` and their consequences on the build?

Reply:
Yes, this come later (still TODO)